### PR TITLE
MONGOCRYPT-284 Wrapped new api in the Java binding

### PR DESCRIPTION
**Pull request by rozza**
Wrapped:
* mongocrypt_setopt_kms_providers - to support new KMS provider options.
* mongocrypt_ctx_setopt_key_encryption_key - to support new data key options.
* mongocrypt_setopt_crypto_hook_sign_rsaes_pkcs1_v1_5 - to support signature callback for GCP.

MONGOCRYPT-284

### CODE REVIEW

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (jyemin) [inline bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:238]: The spec says:

> This [mongocrypt_ctx_setopt_key_encryption_key] accepts a BSON document exactly matching the data key options, but expects the BSON document to also have a "provider" field set to either "azure" or "gcp". 

I don't see where the provider field has been added.

**Comment by kevinAlbs** [comment]
Regarding the few failing `build-and-test-java` tasks on the patch build, my guess is that the mappings in https://github.com/mongodb/libmongocrypt/blob/master/bindings/java/mongocrypt/build.gradle.kts#L123:L136 are out-of-date (and consequently they're fetching an older libmongocrypt without the new API, resulting the linker error).

E.g. https://evergreen.mongodb.com/task/libmongocrypt_macos_build_and_test_and_upload_patch_ccfc731a75f77b3a0fa498a93c8a08b025e9c3cf_5f802000d6d80a2339321610_20_10_09_08_32_02 uploads to 
https://mciuploads.s3.amazonaws.com/libmongocrypt/macos/master/latest/5f802000d6d80a2339321610/libmongocrypt.tar.gz

The mapping indicates it is looking for "darwin" in the URL.

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (kevinAlbs) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:238]: +1

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:154]: Currently the driver passes on the options as set by the user. I didn't want to add too much logic in the driver / capi bindings.  Initially I coded a `get(Azure|Gcp)KmsProviderOptions` mandating the individual options for each KMS. However, I felt that was over kill and that the validation would be handled by the `mongocrypt_setopt_kms_providers` - However happy to revisit that if we want more validation in Java code.

Once #116 is merged do you want this API to stop calling `mongocrypt_setopt_kms_provider_local` and `mongocrypt_setopt_kms_provider_aws` altogether ?

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:238]: That was a little unclear to me as what we expect from the user, we seem to use the following format in the database:

```
"masterKey": {
    "provider": "azure",
    "keyVaultEndpoint": "my-key-vault.vault.azure.net",
    "keyName": "test-key"
}
```

And I had some expectations that the field was explicit.

I was expecting users to use that format eg:

```
{
"azure": {
    "provider": "azure",
    "keyVaultEndpoint": "my-key-vault.vault.azure.net",
    "keyName": "test-key"
}}
```

I'll update and treat the provider as implicit if not explicitly provided.

```
"azure": {
    "keyVaultEndpoint": "my-key-vault.vault.azure.net",
    "keyName": "test-key"
}
```

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/CAPI.java
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/CAPI.java]: Updated.

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (rozza) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java]: Updated

**Comment by rozza** [comment]
I'll look into evergreen - the mapping is for JNA - its looking to download the macos version and it maps to darwin arch.

I noticed this error for Ubuntu:  

>  Error looking up function 'mongocrypt_setopt_crypto_hook_sign_rsaes_pkcs1_v1_5': /usr/lib/x86_64-linux-gnu/libmongocrypt.so: undefined symbol: mongocrypt_setopt_crypto_hook_sign_rsaes_pkcs1_v1_5

Looks like its picked up libmongocrypt from `/usr/lib/` which looks to be the cause there.

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (kevinAlbs) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java:156]: Sounds good to me. I'll leave it up to you whether you want to consolidate when #116 is merged. The other API will continue to work for backwards compatibility.

**Comment by rozza** [comment]
All java tests now passing.

File: bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java
Comment (jyemin) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java]: @kevinAlbs how should this be tested?

Comment (kevinAlbs) [reply bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java]: libmongocrypt verifies that "provider" is present. If the driver omits it, the prose tests for creating a data key for Azure or GCP will fail.

**Review by jyemin — APPROVED** [review]
LGTM!

**Review by kevinAlbs — APPROVED** [review]

**Pull request merged**
